### ISSUE TITLE
binding variables

### ISSUE THREAD

**Issue by peterritter** [Fytch/ProgramOptions.hxx#8]
hi
Here is another idea. How about being able to 'bind' an option to a variable. This would work expecially well for 'mandatory' options. That way, the value would be set directly in a corresponding variable. Like:

double some_num;
parser["some_num"].bind(some_num);  //mandatory is implicit

for multi you could bind to a vector.

std::vector<double> some_nums;
parser["some_nums"].bind(some_nums);  //multi is implicit

I hope you appreciate my ideas. Let me know if you get tired of them:-)

Peter

**Comment by Fytch** [comment]
Hello

I have recently thought of that too! Luckily this is very easy to implement and I'm going to create a branch for it shortly.
Implicit multi is very sensible but implicit mandatory for bound options? I'm not entirely convinced but perhaps it's just me because I vastly prefer explicit stuff. :) Furthermore, most UNIX commands have none or maybe 1 mandatory options, so making every option with a bound variable mandatory might be disadvantageous. What do you think?

Yes, I really appreciate your feedback/suggestions, keep it coming!

Regards, Josua

**Comment by peterritter** [comment]
>> I vastly prefer explicit stuff.

well, I do too in away - I prefer explicit error messages or exceptions. The main problem with bound variables is how to deal with 'undefined/NA/null'. 
Lets say you do this:

double some_num;
parser["some_num"].bind(some_num); 

What value will some_num have when the option was not given by end user? The best would be for the value to be 'undefined', but this is not possible with type 'double'. Something like type boost::any or boost::variant could maybe handle an 'undefined' state but I have to say that I am using your library only because I don't want dependency on boost, otherwise I'd still be using boost::program_options instead. Having said that, if you are going to reinvent the wheel, you might as well make a better wheel altogether!

My suggestion would be to make a template type like:

po::arg<double> my_arg;                  
parser["my_arg"].bind(my_arg);
//....
if(my_arg.defined()){
     double d =  some_num.get();
}
Without such a wrapper template, you'd have to make such bound options mandatory to avoid having to represent 'null' somehow. 

Using bound variables would also make option declaration simpler, since you would not have to do the .type( po::u32 )  thing. You know I don't like that part so much. 

I'm just thinking aloud here. I think finding the right design takes a lot of consideration. But I definitely think that this whole area of program options could benefit from some radical improvement. Performance does not matter at all, since the code is only run once. So I think the emphasis should be on full feature set, ease of use, and perceived simplicity.  Something you can use that is so straight forward one does not even need a manual. 

Peter

**Comment by mrzv** [comment]
It would be really nice if binding a variable played nicely with its existing value. Ideally, it should be equivalent to implicitly calling `failsafe` with the variable's value. In other words, I'm hoping the help message will show the existing value as the default.
```
double x = 1.0;
parser['var'].bind(x);   // implicitly .failsafe(1.0)
```

As for defined vs undefined. Doesn't `parser["my_arg"].available()` already solve this problem?

All this is to say I'm also interested in this feature.

**Comment by Fytch** [comment]
Binding variables has been implemented in 5bcf23a9735a41536a9573304082d49e6a57da8f.

@mrzv I suppose, instead of `.failsafe` you meant `.fallback`? If so: I'm not entirely convinced that copying the existing value into the fallback memory is a good idea.
1. If the caller of `.bind` forgets to initialize the variable, `.bind` reads from uninitialized memory and thus generates undefined behavior.
2. If you initialize a variable and call `.bind` on it, you're going to read from that variable anyway and not from `option.get()`, so the fallback mechanism is already implicitly given.
3. If I wanted to bind a vector where there are already a lot of elements in it, it'd copy the entire existing vector into the fallback memory.

**Comment by mrzv** [comment]
Yeah, sorry, I did mean `fallback`. I see your points. It's mostly a matter of taste. To me initializing variables and skipping `fallback` is a little more natural and also more concise. I'd see the default values right where the variables are defined. But:

1. I see how it could also result in undefined behavior (although the compiler should warn about it, I believe).
2. As a user, I'd find this kind of implicit fallback a bit dangerous, plus it wouldn't show the default value in the help message.
3. If the default value is a large vector, then it probably should be copied, shown in help, etc. It's rare that command-line options are so huge that extra copying matters.

But, like I said, I understand how you see it, too.

**Comment by Fytch** [comment]
I wrote an example for `.bind` (https://github.com/Fytch/ProgramOptions.hxx/commit/d62b574202453815a8287fdaf768661651b9ec3e) and added a chapter in the readme (https://github.com/Fytch/ProgramOptions.hxx/commit/27b6ec6948cede6c95dccc4e89b32df2e8b565fc).

**Comment by Fytch** [comment]
@mrzv 
I think we misunderstood each other a little. Just to clear up:

Your stance is that the user must initialize the to-be-bound variable and calling `.bind` should copy the variable's value into the option's fallback memory.

My stance is that the user may initialize the to-be-bound variable but calling `.bind` doesn't change the option's fallback memory.

Currently, if the user initializes the to-be-bound variable, passes it to `.bind` and then reads from the variable, the value with which it was initialized is already the fallback value, because the parser doesn't change the variable if the corresponding option didn't appear in the program arguments. You don't actually have to call `.fallback`. The only case where you'd need to still call `.fallback` is when you bind a variable and then after parsing decide that you want to access the value via `parser["...].get()` (which you shouldn't do; why bind then anyway?).

In summary, you can initialize to-be-bound variables and skip calling `.fallback` and still get the fallback behavior if you read from the bound variable (which you'd do anyway). Check out [the bind example](https://github.com/Fytch/ProgramOptions.hxx/blob/master/examples/5%20bind.cxx) to see this in action. (consider the variable `optimization`)

Regarding undefined behavior, overlooking to initialize the to-be-bound variable will result in UB either way. Under your proposal, if you forget to initialize the variable, then pass it to `.bind`, the parser will read it (UB) in order to store a copy in the fallback memory.

As to 2.: The default help printer currently doesn't show default values, but that's actually a good idea. I have to think about it.

**Comment by mrzv** [comment]
Yeah, I see it now. I was misunderstanding the details. Thanks for the explanation.
